// electron-builder/out/targets/targetFactory.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.DIR_TARGET = exports.DEFAULT_TARGET = exports.commonTargets = undefined;
exports.createTargets = createTargets;
exports.createCommonTarget = createCommonTarget;

var _platformPackager;

function _load_platformPackager() {
    return _platformPackager = require("../platformPackager");
}

const commonTargets = exports.commonTargets = ["dir", "zip", "7z", "tar.xz", "tar.lz", "tar.gz", "tar.bz2"];
const DEFAULT_TARGET = exports.DEFAULT_TARGET = "default";
const DIR_TARGET = exports.DIR_TARGET = "dir";
function createTargets(nameToTarget, rawList, outDir, packager, cleanupTasks) {
    const result = [];
    const mapper = (name, factory) => {
        let target = nameToTarget.get(name);
        if (target == null) {
            target = factory(outDir);
            nameToTarget.set(name, target);
        }
        result.push(target);
    };
    const targets = normalizeTargets(rawList == null || rawList.length === 0 ? packager.platformSpecificBuildOptions.target : rawList);
    packager.createTargets(targets == null ? [DEFAULT_TARGET] : targets, mapper, cleanupTasks);
    return result;
}
function normalizeTargets(targets) {
    if (targets == null) {
        return null;
    } else {
        return (Array.isArray(targets) ? targets : [targets]).map(it => it.toLowerCase().trim());
    }
}
function createCommonTarget(target) {
    if (!(commonTargets.indexOf(target) !== -1)) {
        throw new Error(`Unknown target: ${ target }`);
    }
    return new (_platformPackager || _load_platformPackager()).Target(target);
}
